src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        // Очищаем текущее состояние
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        // Добавляем свойства из extraData
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        // Удаляем указанные свойства
+        currentState = { ...currentState }; // Клонируем текущее состояние
+
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    // Добавляем текущее состояние в историю
+    stateHistory.push(currentState);
```

Comment:
[CODE STYLE] - It's better to push a copy of the object to the `stateHistory` array to avoid potential mutations to the states stored inside the history. Use `stateHistory.push({ ...currentState });` instead.